osquery/tables/system/darwin/quarantined_history.cpp

```diff
@@ -0,0 +1,122 @@
+
+// Path (after /Users/foo) where the quarantine events DB will be found
+const std::string kQuarantineEventsDbPath =
+    "Library/Preferences/com.apple.LaunchServices.QuarantineEventsV2";
+
+const int cocoaToUnixEpochOffset = 978307200;
+
+void genQuarantineEventRow(sqlite3_stmt* const stmt,
+                           std::string username,
+                           Row& r) {
+  for (int i = 0; i < sqlite3_column_count(stmt); i++) {
+    auto column_name = std::string(sqlite3_column_name(stmt, i));
+    auto column_type = sqlite3_column_type(stmt, i);
+    if (column_type == SQLITE_TEXT) {
+      auto value = sqlite3_column_text(stmt, i);
+      if (value != nullptr) {
+        r[column_name] = std::string(reinterpret_cast<const char*>(value));
+      }
+    } else if (column_type == SQLITE_FLOAT) {
+      auto value = sqlite3_column_double(stmt, i);
+      // convert to UNIX epoch
+      if (column_name == "timestamp") {
+        auto timestamp = value + cocoaToUnixEpochOffset;
+        r[column_name] = DOUBLE(timestamp);
+      } else {
+        r[column_name] = DOUBLE(value);
+      }
+    } else if (column_type == SQLITE_INTEGER) {
+      auto value = sqlite3_column_int(stmt, i);
+      r[column_name] = INTEGER(value);
+    }
+  }
+  r["username"] = TEXT(username);
+}
+
+void genQuarantineHistoryItems(const fs::path& qepath,
+                               std::string username,
+                               QueryData& results) {
+  sqlite3* db = nullptr;
+
+  try {
+    if (!fs::exists(qepath)) {
+      return;
+    }
+  } catch (const boost::filesystem::filesystem_error) {
+    return;
+  }
+
+  auto rc = sqlite3_open_v2(
+      qepath.c_str(),
+      &db,
+      (SQLITE_OPEN_READONLY | SQLITE_OPEN_PRIVATECACHE | SQLITE_OPEN_NOMUTEX),
+      nullptr);
+  if (rc != SQLITE_OK || db == nullptr) {
+    VLOG(1) << "Cannot open Quarantine Events DB: " << rc << " "
+            << getStringForSQLiteReturnCode(rc);
+    if (db != nullptr) {
+      sqlite3_close(db);
+    }
+  }
+
+  const std::string query =
+      "SELECT LSQuarantineEventIdentifier as id, LSQuarantineAgentName as "
+      "agent_name, LSQuarantineAgentBundleIdentifier as "
+      "agent_bundle_identifier, LSQuarantineTypeNumber as type,  "
+      "LSQuarantineDataURLString as data_url,LSQuarantineOriginURLString as "
+      "origin_url, LSQuarantineSenderName as sender_name, "
+      "LSQuarantineSenderAddress as sender_address, LSQuarantineTimeStamp as "
+      "timestamp from LSQuarantineEvent";
+  sqlite3_stmt* stmt = nullptr;
+  rc = sqlite3_prepare_v2(db, query.c_str(), -1, &stmt, nullptr);
+  while ((sqlite3_step(stmt)) == SQLITE_ROW) {
+    Row r;
+    genQuarantineEventRow(stmt, username, r);
+    results.push_back(r);
+  }
+
+  // Clean up.
+  sqlite3_finalize(stmt);
+  sqlite3_close(db);
+}
+
+QueryData genQuarantineHistory(QueryContext& context) {
+  QueryData results;
+
+  // Get the quarantine events for each user.
+  auto users = SQL::selectAllFrom("users");
```

Comment:
Can you use this pattern: https://github.com/facebook/osquery/blob/master/osquery/tables/system/posix/known_hosts.cpp#L61

The problem is, that pattern works for `uid`: https://github.com/facebook/osquery/blob/master/specs/posix/known_hosts.table#L5

Making it work for `username` (to bubble up the performance) you would need to edit the logic here: https://github.com/facebook/osquery/blob/master/osquery/tables/system/darwin/user_groups.mm#L125

Right now, `username` or no `username` in the WHERE clause means enumerate everything, similar to what you're doing here now. Note that on Linux this pattern does result in a performance win because the [case is handled](https://github.com/facebook/osquery/blob/master/osquery/tables/system/linux/users.cpp#L66).